### PR TITLE
Add put, delete, and read repeated entity methods to ApplicantData.

### PR DESCRIPTION
**Pull request by kendrickt**
### Description
`ApplicantData` now has special put, delete, and read methods for repeated entities. This unblocks creating an `ApplicantQuestion` for the repeater question type, which is needed to create the admin view to create a new repeater question.

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### Issue(s)

### CODE REVIEW

File: universal-application-tool-0.0.1/app/services/Path.java [hunk 1]
```diff
@@ -127,21 +127,27 @@ public Path withoutArrayReference() {
...
+      try {
+        return Integer.valueOf(matcher.group(ARRAY_INDEX_GROUP));
+      } catch (NumberFormatException e) {
+        throw new IllegalStateException(
+            String.format("This path %s does not reference a repeated entity element.", this), e);
+      }
     }
     throw new IllegalStateException(
-        String.format("This path %s does not reference an array element.", this));
+        String.format("This path %s does not reference a repeated entity collection.", this));
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/Path.java:144]: Suggestion (blocking): the same error is thrown in two branches of this code. Could be re-written more simply as

```java
    Matcher matcher = ARRAY_INDEX_REGEX.matcher(keyName());

    try {
      return Integer.valueOf(matcher.group(ARRAY_INDEX_GROUP));
    } catch (NumberFormatException e) {
      throw new IllegalStateException(
          String.format("This path %s does not reference a repeated entity element.", this), e);
    }
```

File: universal-application-tool-0.0.1/app/services/Path.java [hunk 2]
```diff
@@ -127,21 +127,27 @@ public Path withoutArrayReference() {
   /**
    * Return the index of the array element this path is referencing.
    *
-   * <p>For paths to non-array elements, {@code IllegalStateException is thrown}.
+   * <p>For paths to non repeated entity collections, or paths to an entire repeated entity
+   * collection without specifying an index, {@code IllegalStateException} is thrown.
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/Path.java:131]: Suggestion (blocking): please clarify that this method will return the *last* array index in the path.

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java [hunk 3]
```diff
@@ -118,6 +119,36 @@ public void putList(Path path, ImmutableList<String> value) {
...
+
+  /**
+   * Deletes the entire repeated entity at the index of the path provided.
+   *
+   * @param path a path to a repeated entities list
+   * @param index the index to delete
+   */
+  public void deleteRepeatedEntity(Path path, int index) {
+    this.jsonData.delete(path.atIndex(index).toString());
+  }
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:150]: Question (blocking): what user behavior will use this code path?

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java [hunk 4]
```diff
@@ -23,6 +23,7 @@
 public class ApplicantData {
   private static final String EMPTY_APPLICANT_DATA_JSON = "{ \"applicant\": {}, \"metadata\": {} }";
   private static final Locale DEFAULT_LOCALE = Locale.US;
+  static final String REPEATED_ENTITY_NAME_KEY = "entity_name";
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:26]: Suggestion (blocking): this string already in appears in the codebase [here](https://github.com/seattle-uat/civiform/blob/47d7145e80098aa630fc8178396f61f34bcf410f/universal-application-tool-0.0.1/app/services/question/RepeaterQuestionDefinition.java#L108). Please create a shared constant to reference in both usages.

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java (hunk 3 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:150]: In the repeater question, applicants will be able to delete a repeated entity by clicking a trash can button.

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java [hunk 5]
```diff
@@ -118,6 +119,36 @@ public void putList(Path path, ImmutableList<String> value) {
     }
   }
 
+  /**
+   * Puts the names of the repeated entities at the path. Each element in the JSON array at the path
+   * is a JSON object with an "entity_name" scalar and possibly other nested answers to questions or
```
Comment (bion) [inline universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:124]: Suggestion (non-blocking):
```suggestion
   * is a JSON object that has at minimum a property "entity_name" that contains a string value, along with possibly other nested answers to questions or
```

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java (hunk 3 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:150]: That's going to be tricky to provide a good UX for without making it API-driven. Though maybe for just that feature a small API endpoint and some JS wouldn't be so bad.

File: universal-application-tool-0.0.1/app/services/Path.java (hunk 1 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/Path.java]: done! It needs to also catch the `IllegalStateException` for when there is no match at all.

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java (hunk 4 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:26]: good catch! fixed!

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java (hunk 5 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:124]: done

File: universal-application-tool-0.0.1/app/services/Path.java (hunk 2 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/Path.java:131]: done

File: universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java (hunk 3 above)
Comment (bion) [reply universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java:153]: Thinking about this some more, the API-driven approach would end up being a bit heavier than I was thinking last night.

**Review by bion — APPROVED** [review]

File: universal-application-tool-0.0.1/test/services/applicant/ApplicantDataTest.java [hunk 6]
```diff
@@ -216,6 +216,54 @@ public void putList_writesNullIfListIsEmpty() {
...
+    assertThat(data.asJsonString())
+        .isEqualTo(
+            "{\"applicant\":{\"children\":[{},{\"pets\":[{\"entity_name\":\"bubbles\"},{\"entity_name\":\"luna\"},{\"entity_name\":\"taco\"}]}]},\"metadata\":{}}");
+  }
+
+  @Test
+  public void putRepeatedEntities_withPrexistingData() {
+    ApplicantData data =
+        new ApplicantData(
+            "{\"applicant\":{\"children\":[{},{\"entity_name\":\"an old name\",\"pets\":["
```
Comment (carolinedanzi) [inline universal-application-tool-0.0.1/test/services/applicant/ApplicantDataTest.java:236]: question: why is there no index 0 entity?

Comment (carolinedanzi) [reply universal-application-tool-0.0.1/test/services/applicant/ApplicantDataTest.java:236]: ah so this renames the existing entity_names? What if we want to add more?

**Review by carolinedanzi — COMMENTED** [review]
Exciting to see this coming together!

**Review by carolinedanzi — APPROVED** [review]

File: universal-application-tool-0.0.1/test/services/applicant/ApplicantDataTest.java (hunk 6 above)
Comment (kendrickt) [reply universal-application-tool-0.0.1/test/services/applicant/ApplicantDataTest.java:236]: this test checks that renames work without changing other parts of the repeated entity (e.g. pets).  The other test tests adding new ones.

File: universal-application-tool-0.0.1/app/services/Path.java [hunk 7]
```diff
@@ -118,7 +118,7 @@ public boolean isArrayElement() {
    * Returns a path with a trailing array element reference stripped away. For example, {@code
    * applicant.children[2]} would return a path to {@code applicant.children}.
    *
-   * <p>For paths to non-array elements, {@code IllegalStateException is thrown}.
+   * <p>For paths to non repeated entity collections, {@code IllegalStateException} is thrown.
```
[the remark this answers is not in the document]
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/Path.java:121]: I _think_ these methods are really just for repeated entities. Is there a situation where we'd need to manipulate indices on multi-select question answers?

[the remark this answers is not in the document]
Comment (kendrickt) [reply universal-application-tool-0.0.1/app/services/Path.java:121]: fixed!

**Pull request merged**